File: 0763-partition-labels/0763-partition-labels.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        result=[]
        counter=collections.Counter(s)

        change = set()
        length = 0
        for i,current in enumerate(s):
            length += 1
            change.add(current)
            counter[current] -= 1 
            
            if counter[current]==0:
                change.remove(current)
                if len(change)==0:
                    result.append(length)
                    length=0
        return result
        

File: 0763-partition-labels/0763-partition-labels.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution {
    public List<Integer> partitionLabels(String s) {
        List<Integer> result = new ArrayList<>();
        Map<Character, Integer> lastIndexMap = new HashMap<>();
        
        for (int i = 0; i < s.length(); i++) {
            lastIndexMap.put(s.charAt(i), i);
        }
        int start = 0;
        int end = 0;
        for (int i = 0; i < s.length(); i++) {
            end = Math.max(end, lastIndexMap.get(s.charAt(i)));

            if (i == end) {
                result.add(end - start + 1);
                start = end + 1;
            }
        }

        return result;
    }
}
